tools/cfn2ts/lib/codegen.ts

```diff
@@ -576,41 +576,95 @@ export default class CodeGenerator {
       }
 
       public visitList(itemType: genspec.CodeName): string {
-        const arg = `prop${this.depth}`;
         return itemType.className === 'string'
           // an array of strings is a special case,
           // because it might need to be encoded as a Token directly
           // (and not an array of tokens), for example,
           // when a Ref expression references a parameter of type CommaDelimitedList
           ? `${CFN_PARSE}.FromCloudFormation.getStringArray(${this.baseExpression})`
-          : `${CFN_PARSE}.FromCloudFormation.getArray(${this.baseExpression}, (${arg}: any) => ` +
-            `${this.deeperCopy(arg).visitAtom(itemType)})`;
+          : `${CFN_PARSE}.FromCloudFormation.getArray(${this.baseExpression}, ${this.renderVisitAtomLambdaExpression(itemType)})`;
       }
 
       public visitMap(itemType: genspec.CodeName): string {
-        const arg = `prop${this.depth}`;
-        return `${CFN_PARSE}.FromCloudFormation.getMap(${this.baseExpression}, (${arg}: any) => ` +
-          `${this.deeperCopy(arg).visitAtom(itemType)})`;
+        return `${CFN_PARSE}.FromCloudFormation.getMap(${this.baseExpression}, ${this.renderVisitAtomLambdaExpression(itemType)})`;
       }
 
-      public visitAtomUnion(_types: genspec.CodeName[]): string {
-        return this.baseExpression;
+      public visitAtomUnion(types: genspec.CodeName[]): string {
+        const validatorNames = types.map(type => genspec.validatorName(type).fqn).join(', ');
+
+        const mappers = types
+          .map(type => this.renderVisitAtomLambdaExpression(type))
+          .join(', ');
+
+        return `${CFN_PARSE}.FromCloudFormation.getTypeUnion([${validatorNames}], [${mappers}], ${this.baseExpression})`;
       }
 
-      public visitListOrAtom(_scalarTypes: genspec.CodeName[], _itemTypes: genspec.CodeName[]): any {
-        return this.baseExpression;
+      public visitUnionList(itemTypes: genspec.CodeName[]): string {
+        const validatorNames = itemTypes.map(type => genspec.validatorName(type).fqn).join(', ');
+
+        const arg = `prop${this.depth}`;
+        // we need a deeper copy here, as 'mappers' is used in a lambda expression below
+        const deeperVisitor = this.deeperCopy();
```

Comment:
Again, this has the goal of JUST increasing `depth`, which is not super clear to me.